File: src/main/java/eu/mshade/enderman/packet/play/MinecraftPacketOutEntityProperties.java
package eu.mshade.enderman.packet.play;

import eu.mshade.enderframe.metadata.attribute.AttributeProperty;
import eu.mshade.enderframe.protocol.MinecraftPacketOut;
import eu.mshade.enderframe.protocol.MinecraftByteBuf;

import java.util.List;

public class MinecraftPacketOutEntityProperties implements MinecraftPacketOut {
    private final int id;
    private final List<AttributeProperty> attributeProperties;

    public MinecraftPacketOutEntityProperties(int id, List<AttributeProperty> attributeProperties) {
        this.id = id;
        this.attributeProperties = attributeProperties;
    }

    @Override
    public void serialize(MinecraftByteBuf minecraftByteBuf) {
        minecraftByteBuf.writeVarInt(id);
        minecraftByteBuf.writeInt(attributeProperties.size());
        attributeProperties.forEach(attributeProperty -> {
            minecraftByteBuf.writeString(attributeProperty.getAttribute().getName());
            minecraftByteBuf.writeDouble(attributeProperty.getValue());
            minecraftByteBuf.writeVarInt(0);
        });
    }
}

File: src/main/java/eu/mshade/enderman/packet/play/entity/MinecraftPacketOutEntityProperties.kt
package eu.mshade.enderman.packet.play.entity

import eu.mshade.enderframe.attribute.AttributeKey
import eu.mshade.enderframe.attribute.AttributeModifier
import eu.mshade.enderframe.attribute.AttributeProperty
import eu.mshade.enderframe.entity.Entity
import eu.mshade.enderframe.protocol.MinecraftByteBuf
import eu.mshade.enderframe.protocol.MinecraftPacketOut
import java.util.function.Consumer

class MinecraftPacketOutEntityProperties(
    private val entity: Entity,
    private val attributeProperties: List<EndermanAttributeProperty>
) : MinecraftPacketOut {

    override fun serialize(minecraftByteBuf: MinecraftByteBuf) {
        minecraftByteBuf.writeVarInt(entity.getEntityId())
        minecraftByteBuf.writeInt(attributeProperties.size)

        attributeProperties.forEach {attributeProperty ->

            minecraftByteBuf.writeString(attributeProperty.attribute)
            minecraftByteBuf.writeDouble(attributeProperty.value)
            minecraftByteBuf.writeVarInt(attributeProperty.attributeModifiers.size)

            attributeProperty.attributeModifiers.forEach{modifier ->
                minecraftByteBuf.writeUUID(modifier.uuid)
                minecraftByteBuf.writeDouble(modifier.amount)
                minecraftByteBuf.writeByte(modifier.attributeOperation.ordinal)
            }

        }
    }

}


data class EndermanAttributeProperty(val attribute: String, val value: Double, val attributeModifiers: List<AttributeModifier>)

